### PR TITLE
Add docs on Git source type for Function

### PR DESCRIPTION
**Pull request by kazydek**
**Description**

Changes proposed in this pull request:

I described an alternative Git source of Function's code and dependencies:

 - Added a general Git Source type doc describing the idea
 - Updated the Function CR doc with missing status conditions and new fields for the Git source
 - Added a new GitRepository CR document describing the new custom resource 
 - Updated the doc on validation and defaulting webhook to reflect also new GitRepository CR

### CODE REVIEW

**Review by m00g3n — APPROVED** [review]

File: docs/serverless/06-01-function-cr.md [hunk 1]
```diff
@@ -70,29 +70,47 @@ spec:
...
+| **spec.resources.requests.cpu**          |       No       | Specifies the number of CPUs requested by the Function's Pod to operate.       |
+| **spec.resources.requests.memory**       |       No       | Specifies the amount of memory requested by the Function's Pod to operate.               |
+| **spec.runtime**                         |       No       | Specifies the runtime of the Function. The available values are `nodejs12`, `nodejs10`, and `python38`. It is set to `nodejs12` unless specified otherwise.  |
+| **spec.type**                          |      No       | Defines that you use a Git repository as the source of Function's code and dependencies. It must be set to `git`. |
+| **spec.source**                          |      Yes       | Provides the Function's full source code or the name of the Git directory in which the code and dependencies are stored.     |
+| **spec.baseDir**                          |      No       | Specifies the relative path to the Git directory containing the source code from which the Function will be built​. |
```
Comment (klaudiagrz) [inline docs/serverless/06-01-function-cr.md:108]:
```suggestion
| **spec.baseDir**                          |      No       | Specifies the relative path to the Git directory that contains the source code from which the Function will be built​. |
```

File: docs/serverless/06-02-gitrepository-cr.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+metadata:
+  name: sample-with-auth
+spec:
+  url: "git@github.com:sample-organization/sample-repo.git"
+  auth:
+    type: key
+    secretName: kyma-git-creds
+```
+
+## Custom resource properties
```
Comment (klaudiagrz) [inline docs/serverless/06-02-gitrepository-cr.md:28]:
```suggestion
## Custom resource parameters
```

File: docs/serverless/06-02-gitrepository-cr.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+This table lists all the possible properties of a given resource together with their descriptions:
```
Comment (klaudiagrz) [inline docs/serverless/06-02-gitrepository-cr.md:30]:
```suggestion
This table lists all the possible parameters of a given resource together with their descriptions:
```

File: docs/serverless/06-02-gitrepository-cr.md (hunk 2 above)
Comment (klaudiagrz) [reply docs/serverless/06-02-gitrepository-cr.md:28]: See the [template](https://github.com/kyma-project/community/blob/master/guidelines/templates/resources/custom-resource.md#custom-resource-parameters) for reference.

File: docs/serverless/06-02-gitrepository-cr.md (hunk 3 above)
Comment (klaudiagrz) [reply docs/serverless/06-02-gitrepository-cr.md:30]: As above.

File: docs/serverless/06-02-gitrepository-cr.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+| **spec.auth** | No | Specifies that you must authenticate to the Git repository. |
+| **spec.auth.type** | No | Defines if you must authenticate to the repository with a password or token (`basic`), or an SSH key (`key`). This parameter is required if you provide **spec.auth**. |
+| **spec.auth.secretName** | No | Specifies the name of the Secret with credentials used by the Function Controller to authenticate to the Git repository to fetch the Function's source code and dependencies. This Secret must be stored in the same Namespace as the GitRepository CR. The **spec.auth.secretName** parameter is required if you provide **spec.auth**. |
+
+## Related resources and components
+
+The Function custom resource relies on these Kubernetes resources:
+
+| Resource           | Description                                                                                                  |
```
Comment (klaudiagrz) [inline docs/serverless/06-02-gitrepository-cr.md:43]:
```suggestion
| Custom resource           | Description                                                                                                  |
```

Comment (klaudiagrz) [reply docs/serverless/06-02-gitrepository-cr.md:43]: As above.

File: docs/serverless/06-02-gitrepository-cr.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    secretName: kyma-git-creds
+```
+
+## Custom resource properties
+
+This table lists all the possible properties of a given resource together with their descriptions:
+
+| Parameter | Required | Description |
+|-----------|-------------|---------------|
+| **spec.url** | Yes | Provides the address to the Git repository with the Function's code and dependencies. Depending if the repository is public or private and what authentication method is used to access it, URL must start with the `http(s)`, `git`, or `ssh` prefix, and end with the `.git` suffix.  |
```
Comment (klaudiagrz) [inline docs/serverless/06-02-gitrepository-cr.md:34]:
```suggestion
| **spec.url** | Yes | Provides the address to the Git repository with the Function's code and dependencies. Depending on whether the repository is public or private and what authentication method is used to access it, the URL must start with the `http(s)`, `git`, or `ssh` prefix, and end with the `.git` suffix.  |
```

File: docs/serverless/06-02-gitrepository-cr.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+| **spec.auth** | No | Specifies that you must authenticate to the Git repository. |
+| **spec.auth.type** | No | Defines if you must authenticate to the repository with a password or token (`basic`), or an SSH key (`key`). This parameter is required if you provide **spec.auth**. |
+| **spec.auth.secretName** | No | Specifies the name of the Secret with credentials used by the Function Controller to authenticate to the Git repository to fetch the Function's source code and dependencies. This Secret must be stored in the same Namespace as the GitRepository CR. The **spec.auth.secretName** parameter is required if you provide **spec.auth**. |
```
Comment (klaudiagrz) [inline docs/serverless/06-02-gitrepository-cr.md:37]:
```suggestion
| **spec.auth.secretName** | No | Specifies the name of the Secret with credentials used by the Function Controller to authenticate to the Git repository in order to fetch the Function's source code and dependencies. This Secret must be stored in the same Namespace as the GitRepository CR. The **spec.auth.secretName** parameter is required if you provide **spec.auth**. |
```

File: docs/serverless/06-02-gitrepository-cr.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+## Related resources and components
+
+The Function custom resource relies on these Kubernetes resources:
+
+| Resource           | Description                                                                                                  |
+| ------------------- | ------------------------------------------------------------------------------------------------------------ |
+| [Function](#custom-resource-function)     | Stores the Function's source code and dependencies on a cluster.                                   |                                  |
```
Comment (klaudiagrz) [inline docs/serverless/06-02-gitrepository-cr.md:45]:
```suggestion
| [Function](#custom-resource-function)     | Stores the Function's source code and dependencies on a cluster.                                   | 
```

File: docs/serverless/06-02-gitrepository-cr.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+| Parameter | Required | Description |
+|-----------|-------------|---------------|
+| **spec.url** | Yes | Provides the address to the Git repository with the Function's code and dependencies. Depending if the repository is public or private and what authentication method is used to access it, URL must start with the `http(s)`, `git`, or `ssh` prefix, and end with the `.git` suffix.  |
+| **spec.auth** | No | Specifies that you must authenticate to the Git repository. |
+| **spec.auth.type** | No | Defines if you must authenticate to the repository with a password or token (`basic`), or an SSH key (`key`). This parameter is required if you provide **spec.auth**. |
+| **spec.auth.secretName** | No | Specifies the name of the Secret with credentials used by the Function Controller to authenticate to the Git repository to fetch the Function's source code and dependencies. This Secret must be stored in the same Namespace as the GitRepository CR. The **spec.auth.secretName** parameter is required if you provide **spec.auth**. |
+
+## Related resources and components
+
+The Function custom resource relies on these Kubernetes resources:
```
Comment (klaudiagrz) [inline docs/serverless/06-02-gitrepository-cr.md:41]:
```suggestion
These are the resources related to this CR:
```

File: docs/serverless/03-02-git-source-type.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+- `handler.js` or `handler.py` with Function's code
+- `package.json` or `requirements.txt` with Function's dependencies
+
+The Function custom resource (CR) must contain `type: git` to specify that you use a Git repository for the Function's sources.
+
+To create a Function with the Git source, you must:
+
+1. Create a [GitRepository CR](#custom-resource-git-repository) with details of your Git repository.
+2. Create a Secret CR (optional, only if you must authenticate to the repository).
```
[the remark this answers is not in the document]
Comment (kazydek) [reply docs/serverless/03-02-git-source-type.md:18]: I added a link to k8s docs.

File: docs/serverless/03-02-git-source-type.md [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+3. Create a [Function CR](#custom-resource-function) with your Function definition and references to the Git repository.
+
+You can have various setups for your Function's Git source with different:
+
+- Directory structures
+
+  You can specify the location of your code dependencies with the `baseDir` parameter in the Function CR. For example, use `"/"` if you keep the source files at the root of your repository.
```
[the remark this answers is not in the document]
Comment (kazydek) [reply docs/serverless/03-02-git-source-type.md:25]: No, just a mistake :)

File: docs/serverless/03-02-git-source-type.md [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+- Authentication methods
+
+  You can define with the **spec.auth** parameter in the GitRepository CR that you must authenticate to the repository with a password or token (`basic`), or an SSH key (`key`).
```
[the remark this answers is not in the document]
Comment (kazydek) [reply docs/serverless/03-02-git-source-type.md:29]: I think code fonts do the trick, and there is a link to the GitReposiotry CR above so the user can access the spec for details.

File: docs/serverless/03-02-git-source-type.md [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+- Function's rebuild triggers
+
+  You can use the **reference** parameter in the GitRepository CR to define whether the Function Controller must monitor a given branch or commit in the Git repository and rebuild the Function upon their changes.
```
[the remark this answers is not in the document]
Comment (kazydek) [reply docs/serverless/03-02-git-source-type.md:33]: yup 👍  I changed that to `to` to make it shorter.

**Review by klaudiagrz — APPROVED** [review]
LGTM 🚀

**Pull request merged**